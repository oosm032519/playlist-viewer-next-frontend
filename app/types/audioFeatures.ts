export type Mode = "MAJOR" | "MINOR";

export interface AudioFeatures {
    acousticness: number;
    danceability: number;
    energy: number;
    instrumentalness: number;
    liveness: number;
    speechiness: number;
    valence: number;
    tempo: number;
}

export interface TrackAudioFeatures extends AudioFeatures {
    durationMs: number;
    id: string;
    key: number;
    loudness: number;
    mode: Mode;
    timeSignature: number;
}
