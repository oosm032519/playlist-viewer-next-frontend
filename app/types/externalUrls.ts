// app/types/externalUrls.ts
export interface ExternalUrls {
    externalUrls: {
        spotify: string;
    };
}
