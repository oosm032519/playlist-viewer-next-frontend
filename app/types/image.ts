// app/types/image.ts

/**
 * 画像情報。
 */
export interface Image {
    height: number;
    url: string;
    width: number;
}
