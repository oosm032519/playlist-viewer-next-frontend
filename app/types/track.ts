// C:\Users\IdeaProjects\playlist-viewer-next-frontend\app\types\track.ts
// app/types/track.ts
export interface TrackArtist {
    externalUrls: any;
    name: string;
}

export interface TrackAlbum {
    externalUrls: any;
    name: string;
    images: { url: string }[];
}

export interface Track {
    previewUrl: string | undefined;
    id: string | null | undefined;
    name: string;
    artists: TrackArtist[];
    album: TrackAlbum;
    durationMs: number;
    audioFeatures?: {
        danceability: number;
        energy: number;
        key: number;
        loudness: number;
        mode: number;
        speechiness: number;
        acousticness: number;
        instrumentalness: number;
        liveness: number;
        valence: number;
        tempo: number;
        timeSignature: number;
    };
}
