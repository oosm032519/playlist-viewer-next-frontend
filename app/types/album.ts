// app/types/album.ts
import {ExternalUrls} from '@/app/types/externalUrls'
import {Image} from '@/app/types/image'

export interface Album {
    name: string;
    images: Image[];
    externalUrls: ExternalUrls;
}
