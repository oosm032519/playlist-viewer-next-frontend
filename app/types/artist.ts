// app/types/artist.ts
import {ExternalUrls} from '@/app/types/externalUrls'

export interface Artist {
    name: string;
    externalUrls: ExternalUrls;
}
