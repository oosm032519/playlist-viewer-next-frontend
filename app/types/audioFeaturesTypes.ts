import {Track} from '../types/track';

export interface AudioFeaturesChartProps {
    track: Track;
}

export interface AudioFeatureData {
    feature: string;
    value: number;
}
