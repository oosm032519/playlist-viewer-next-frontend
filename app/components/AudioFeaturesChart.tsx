// app/components/AudioFeaturesChart.tsx

import React from 'react';
import {
    Radar,
    RadarChart,
    PolarGrid,
    PolarAngleAxis,
    PolarRadiusAxis,
    ResponsiveContainer,
} from 'recharts';
import {AudioFeaturesChartProps} from '../types/audioFeaturesTypes';
import {prepareAudioFeaturesData} from '../lib/audioFeaturesUtils';

const AudioFeaturesChart: React.FC<AudioFeaturesChartProps> = ({track}) => {
    const data = prepareAudioFeaturesData(track.audioFeatures);
    
    return (
        <ResponsiveContainer width="100%" height={400}>
            <RadarChart cx="50%" cy="50%" outerRadius="80%" data={data}>
                <PolarGrid/>
                <PolarAngleAxis dataKey="feature"/>
                <PolarRadiusAxis angle={30} domain={[0, 1]}/>
                <Radar
                    name="Audio Features"
                    dataKey="value"
                    stroke="#8884d8"
                    fill="#8884d8"
                    fillOpacity={0.6}
                />
            </RadarChart>
        </ResponsiveContainer>
    );
};

export default AudioFeaturesChart;
