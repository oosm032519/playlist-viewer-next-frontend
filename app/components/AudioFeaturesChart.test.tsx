import '@testing-library/jest-dom';
import React from 'react';
import {render, screen} from '@testing-library/react';
import AudioFeaturesChart from './AudioFeaturesChart';
import {Track} from '../types/track';
import {axe, toHaveNoViolations} from 'jest-axe';
import {expect, it} from '@jest/globals';

expect.extend(toHaveNoViolations);

// モックデータ
const mockTrack: Track = {
    id: '1',
    name: 'Test Track',
    artists: [{name: 'Test Artist', externalUrls: undefined}],
    album: {
        name: 'Test Album',
        images: [{url: 'test-image-url'}],
        externalUrls: undefined,
    },
    audioFeatures: {
        danceability: 0.8,
        energy: 0.6,
        key: 5,
        loudness: -5.5,
        mode: 'major',
        speechiness: 0.1,
        acousticness: 0.2,
        instrumentalness: 0.01,
        liveness: 0.3,
        valence: 0.7,
        tempo: 120,
        timeSignature: 4,
    },
    previewUrl: undefined,
    durationMs: 0,
};

// Rechartsコンポーネントをモック
jest.mock('recharts', () => ({
    ResponsiveContainer: ({children}: { children: React.ReactNode }) => (
        <div data-testid="responsive-container">{children}</div>
    ),
    RadarChart: ({children}: { children: React.ReactNode }) => (
        <div data-testid="radar-chart">{children}</div>
    ),
    PolarGrid: () => <div data-testid="polar-grid"/>,
    PolarAngleAxis: () => <div data-testid="polar-angle-axis"/>,
    PolarRadiusAxis: () => <div data-testid="polar-radius-axis"/>,
    Radar: ({dataKey}: { dataKey: string }) => (
        <div data-testid="radar" data-datakey={dataKey}/>
    ),
}));

describe('AudioFeaturesChart', () => {
    it('チャートと必要なコンポーネントをすべてレンダリングする', () => {
        render(<AudioFeaturesChart track={mockTrack}/>);
        
        expect(screen.getByTestId('responsive-container')).toBeInTheDocument();
        expect(screen.getByTestId('radar-chart')).toBeInTheDocument();
        expect(screen.getByTestId('polar-grid')).toBeInTheDocument();
        expect(screen.getByTestId('polar-angle-axis')).toBeInTheDocument();
        expect(screen.getByTestId('polar-radius-axis')).toBeInTheDocument();
        expect(screen.getByTestId('radar')).toBeInTheDocument();
    });
    
    it('undefined の audioFeatures を適切に処理する', () => {
        const trackWithoutAudioFeatures: Track = {...mockTrack, audioFeatures: undefined};
        render(<AudioFeaturesChart track={trackWithoutAudioFeatures}/>);
        
        expect(screen.getByTestId('responsive-container')).toBeInTheDocument();
        expect(screen.getByTestId('radar-chart')).toBeInTheDocument();
    });
    
    it('Radar コンポーネントに正しいデータを渡す', () => {
        render(<AudioFeaturesChart track={mockTrack}/>);
        
        const radarElement = screen.getByTestId('radar');
        expect(radarElement).toHaveAttribute('data-datakey', 'value');
    });
    
    it('部分的な audioFeatures データを処理する', () => {
        const partialAudioFeatures: Partial<Track['audioFeatures']> = {
            danceability: 0.8,
            energy: 0.6,
        };
        const trackWithPartialAudioFeatures: Track = {
            ...mockTrack,
            audioFeatures: partialAudioFeatures as Track['audioFeatures'],
        };
        render(<AudioFeaturesChart track={trackWithPartialAudioFeatures}/>);
        
        expect(screen.getByTestId('responsive-container')).toBeInTheDocument();
        expect(screen.getByTestId('radar')).toBeInTheDocument();
    });
    
    it('スナップショットと一致する', () => {
        const {container} = render(<AudioFeaturesChart track={mockTrack}/>);
        expect(container).toMatchSnapshot();
    });
    
    it('アクセシビリティに問題がない', async () => {
        const {container} = render(<AudioFeaturesChart track={mockTrack}/>);
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});
