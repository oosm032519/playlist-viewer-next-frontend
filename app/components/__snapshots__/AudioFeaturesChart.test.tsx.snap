// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioFeaturesChart matches snapshot 1`] = `
<div>
  <div
    data-testid="responsive-container"
  >
    <div
      data-testid="radar-chart"
    >
      <div
        data-testid="polar-grid"
      />
      <div
        data-testid="polar-angle-axis"
      />
      <div
        data-testid="polar-radius-axis"
      />
      <div
        data-datakey="value"
        data-testid="radar"
      />
    </div>
  </div>
</div>
`;

exports[`AudioFeaturesChart スナップショットと一致する 1`] = `
<div>
  <div
    data-testid="responsive-container"
  >
    <div
      data-testid="radar-chart"
    >
      <div
        data-testid="polar-grid"
      />
      <div
        data-testid="polar-angle-axis"
      />
      <div
        data-testid="polar-radius-axis"
      />
      <div
        data-datakey="value"
        data-testid="radar"
      />
    </div>
  </div>
</div>
`;
